htmresearch/algorithms/apical_tiebreak_temporal_memory.py

```diff
@@ -475,6 +489,56 @@ def _calculateSegmentActivity(connections, activeInput, connectedPermanence,
             matchingSegments,
             potentialOverlaps)
 
+  @staticmethod
+  def _calculateBasalSegmentActivity(connections, activeInput, connectedPermanence,
+                                activationThreshold, minThreshold, reducedThresholdBasal, apicallyActiveCells=()):
+    """
+    Calculate the active and matching basal segments for this timestep.
+
+    The difference with _calculateApicalSegmentActivity is that cells
+    with active apical segments have a lower activation threshold for
+    their basal segments (set by reducedThresholdBasal parameter).
+
+    @param connections (SparseMatrixConnections)
+    @param activeInput (numpy array)
+
+    @return (tuple)
+    - activeSegments (numpy array)
+      Dendrite segments with enough active connected synapses to cause a
+      dendritic spike
+
+    - matchingSegments (numpy array)
+      Dendrite segments with enough active potential synapses to be selected for
+      learning in a bursting column
+
+    - potentialOverlaps (numpy array)
+      The number of active potential synapses for each segment.
+      Includes counts for active, matching, and nonmatching segments.
+    """
+    # Active apical segments lower the activation threshold for basal (lateral) segments
+    overlaps = connections.computeActivity(activeInput, connectedPermanence)
+    outrightActiveSegments = np.flatnonzero(overlaps >= activationThreshold)
+    if reducedThresholdBasal != activationThreshold:
+        conditionallyActiveSegments = np.flatnonzero((overlaps < activationThreshold)
+                                        & (overlaps >= reducedThresholdBasal))
+        cellsOfCASegments = connections.mapSegmentsToCells(conditionallyActiveSegments)
+        # apically active segments are condit. active segments from apically active cells
+        apicallyActiveSegments = conditionallyActiveSegments[np.in1d(cellsOfCASegments,
```

Comment:
I actually added a `filterSegmentsByCell` method that will be useful here:

~~~python
apicallyActiveSegments  = connections.filterSegmentsByCell(
  conditionallyActiveSegments, apicallyActiveCells)
~~~